File: Hyonsok Na/week2/2022-06-14-Maximum-Product-Subarray.java
//used a nested loop :(
class Solution {
    public int maxProduct(int[] nums) {
        int max = 0;
        if(nums.length == 1) return nums[0];
        for(int i = 0; i<nums.length-1; i++) {
            int prev = nums[i];
            for(int j = i+1; j<nums.length; j++) {
                prev = prev*nums[j];
                if(prev>max) max = prev;
                if(nums[j]>max) max = nums[j];
            }
            if(nums[i]>max) max=nums[i];
        }
        return max;
    }
}

File: MinjiJu/2022-summer/2022-06-14-maximum-product-subarray.cpp
// take 2
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        int res=nums[0], left=0, right=0;
        
        for(int i=0; i<n; i++){
            left = (left ? left : 1)*nums[i];           // product from left
            right = (right ? right : 1)*nums[n-1-i];    // product from right
            res = max(res, max(left,right));            // compare and update max
        }
        return res;
    }
};

// init brute force approach
// maximum time limit exceeded

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        
        int prod = nums.at(0);
        
        for(int i=0; i<nums.size(); i++){
            int temp=1;
            for(int j=i; j<nums.size(); j++){
                
                temp*=nums.at(j);
                prod = max(prod,temp);
            }
        }
        return prod;
    }
};